std/experimental/allocator/package.d

```diff
@@ -1123,48 +1123,107 @@ allocated with the same allocator.
 */
 void dispose(A, T)(auto ref A alloc, T* p)
 {
+    if (!p) return;
     static if (hasElaborateDestructor!T)
-    {
-        destroy(*p);
-    }
+        p.__dtor;
     alloc.deallocate((cast(void*)p)[0 .. T.sizeof]);
 }
 
 /// Ditto
 void dispose(A, T)(auto ref A alloc, T p)
-if (is(T == class) || is(T == interface))
+if (is(T == class) && T.stringof != Object.stringof)
+{
+    if (!p) return;
+    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
+    static if (__traits(hasMember, T, "__dtor"))
+        p.__dtor;
+    alloc.deallocate(support);
+}
+
+/// Ditto
+void dispose(A, T, bool isObject = false)(auto ref A alloc, T p)
+if (is(T == class) && T.stringof == Object.stringof)
 {
+    // the content of this overload can't be in the previous because when the
+    // static type is not known (e.g after cast from interface to Object) dispose()
+    // can't be @nogc since the dtor is a delegate determined at runtime. isObject
+    // indicates if the static type is really "Object", e.g "alloc.make!Object"
     if (!p) return;
-    static if (is(T == interface))
+    static if (!isObject)
     {
-        version(Windows)
+        TypeInfo_Class tic = cast(TypeInfo_Class) typeid(p);
+        auto support = (cast(void*) p)[0 .. tic.initializer.length];
+
+        void* dtorPtr = tic.destructor;
+        while (!dtorPtr && tic.base)
         {
-            import core.sys.windows.unknwn;
-            static assert(!is(T: IUnknown), "COM interfaces can't be destroyed in "
-                ~ __PRETTY_FUNCTION__);
+            tic = tic.base;
+            dtorPtr = tic.destructor;
+        }
+        if (dtorPtr)
+        {
+            void delegate() dtor;
+            dtor.funcptr = cast(void function()) dtorPtr;
+            dtor.ptr = cast(void*) p;
+            dtor();
         }
-        auto ob = cast(Object) p;
     }
-    else
-        alias ob = p;
-    auto support = (cast(void*) ob)[0 .. typeid(ob).initializer.length];
-    destroy(p);
     alloc.deallocate(support);
 }
 
 /// Ditto
+void dispose(A, T)(auto ref A alloc, T p)
+if (is(T == interface))
+{
+    if (!p) return;
+    version(Windows)
+    {
+        import core.sys.windows.unknwn: IUnknown;
+        static assert(!is(T: IUnknown), "COM interfaces can't be destroyed in "
+            ~ __PRETTY_FUNCTION__);
+    }
+    static if (__traits(allMembers, T).length)
+    {
+        foreach (ov; __traits(getOverloads, T, __traits(allMembers, T)[0]))
+        static assert(functionLinkage!ov != "C++", "C++ interfaces can't be destroyed in "
+            ~ __PRETTY_FUNCTION__);
+    }
+    dispose(alloc, cast(Object) p);
+}
+
+/// Ditto
 void dispose(A, T)(auto ref A alloc, T[] array)
 {
     static if (hasElaborateDestructor!(typeof(array[0])))
```

Comment:
Otherwise the overload that handles the `@nogc` destructor would look like this

```D
void dispose(A, T)(auto ref A alloc, T p)
if (is(T == class) && T.stringof != Object.stringof)
{
    if (!p) return;
    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
    static if (__traits(hasMember, T, "__dtor"))
        p.__dtor;
    else
    {
        TypeInfo_Class tic = cast(TypeInfo_Class) typeid(p);
        void* dtorPtr = tic.destructor;
        while (!dtorPtr && tic.base)
        {
            tic = tic.base;
            dtorPtr = tic.destructor;
        }
        if (tic && tic.destructor)
        {
            dispose(alloc, cast(Object) p);
            return;
        }
    }
    alloc.deallocate(support);
} 
```

not good.